packages/rollup-plugin/src/lib.ts

```diff
@@ -0,0 +1,121 @@
+import { cssFileFilter } from '@vanilla-extract/integration';
+import MagicString, { Bundle as MagicStringBundle } from 'magic-string';
+import type { ModuleInfo, PluginContext } from 'rollup';
+
+/** Generate a CSS bundle from Rollup context */
+export function generateCssBundle({
+  getModuleIds,
+  getModuleInfo,
+  warn,
+}: Pick<PluginContext, 'getModuleIds' | 'getModuleInfo' | 'warn'>): {
+  bundle: MagicStringBundle;
+  extractedCssIds: Set<string>;
+} {
+  const cssBundle = new MagicStringBundle();
+  const extractedCssIds = new Set<string>();
+
+  // 1. identify CSS files to bundle
+  const cssFiles: Record<string, ImportChain> = {};
+  for (const id of getModuleIds()) {
+    if (cssFileFilter.test(id)) {
+      cssFiles[id] = buildImportChain(id, { getModuleInfo, warn });
+    }
+  }
+
+  // 2. build bundle from import order
+  for (const id of sortModules(cssFiles)) {
+    const { importedIdResolutions } = getModuleInfo(id) ?? {};
+    for (const resolution of importedIdResolutions ?? []) {
+      if (resolution.meta.css) {
+        extractedCssIds.add(resolution.id);
+        cssBundle.addSource({
+          filename: resolution.id,
+          content: new MagicString(resolution.meta.css),
+        });
+      }
+    }
+  }
+
+  return { bundle: cssBundle, extractedCssIds };
+}
+
+/** [id, order] tuple meant for ordering imports */
+export type ImportChain = [string, number][];
+
+/** Trace a file back through its importers, building an ordered list */
+export function buildImportChain(
+  id: string,
+  { getModuleInfo, warn }: Pick<PluginContext, 'getModuleInfo' | 'warn'>,
+): ImportChain {
+  let mod: ModuleInfo = getModuleInfo(id)!;
```

Comment:
Would prefer to not use `!` here and on line 68 if possible. I find `!` usually obfuscates type-issues and a lack of proper error/falsiness handling.